src/metabase/integrations/slack.clj

```diff
@@ -24,75 +23,130 @@
+                  (get-in request [:form-params :token])
+                  (slack-token))]
+    (when token
+      (let [url     (str slack-api-base-url "/" (name endpoint))
+            _       (log/trace "Slack API request: %s %s" (pr-str url) (pr-str request))
+            request (merge-with merge
+                      {:query-params   {:token token}
+                       :as             :stream
+                       :conn-timeout   1000
+                       :socket-timeout 1000}
+                      request)]
+        (try
+          (handle-response (request-fn url request))
+          (catch Throwable e
+            (throw (ex-info (.getMessage e) (merge (ex-data e) {:url url, :request request}) e))))))))
+
+(defn GET
   "Make a GET request to the Slack API."
-  (partial do-slack-request http/get  :query-params))
+  [endpoint & {:as query-params}]
+  (do-slack-request http/get endpoint {:query-params query-params}))
 
-(def ^{:arglists '([endpoint & {:as params}]), :style/indent 1}
-  POST
+(defn POST
   "Make a POST request to the Slack API."
-  (partial do-slack-request http/post :form-params))
-
-(def ^{:arglists '([& {:as args}])} channels-list
-  "Calls Slack api `channels.list` function and returns the list of available channels."
-  (comp :channels (partial GET :channels.list, :exclude_archived true, :exclude_members true)))
-
-(def ^{:arglists '([& {:as args}])} users-list
-  "Calls Slack api `users.list` function and returns the list of available users."
-  (comp :members (partial GET :users.list)))
-
-(def ^:private ^String channel-missing-msg
-  (str "Slack channel named `metabase_files` is missing! Please create the channel in order to complete "
-       "the Slack integration. The channel is used for storing graphs that are included in pulses and "
-       "MetaBot answers."))
-
-(defn- maybe-get-files-channel
-  "Return the `metabase_files channel (as a map) if it exists."
-  []
-  (some (fn [channel] (when (= (:name channel) files-channel-name)
-                        channel))
-        (channels-list :exclude_archived false)))
+  [endpoint body]
+  (do-slack-request http/post endpoint {:form-params body}))
+
+(defn- next-cursor
+  "Get a cursor for the next page of results in a Slack API response, if one exists."
+  [response]
+  (not-empty (get-in response [:response_metadata :next_cursor])))
+
+(def ^:private max-list-results
+  "Absolute maximum number of results to fetch from Slack API list endpoints. To prevent unbounded pagination of results."
+  1000)
+
+(defn- paged-list-request
+  "Make a GET request to a Slack API list `endpoint`, returning a sequence of objects returned by the top level
+  `results-key` in the response. If additional pages of results exist, fetches those lazily, up to a total of
+  `max-list-results`."
+  [endpoint results-key params]
+  (let [response (m/mapply GET endpoint params)]
+    (when (seq response)
+      (take
+       max-list-results
```

Comment:
@camsaul I'm not a Clojure-head so might be reading this wrong, but this looks like it'll only ever take the first 1,000 list items even after fetching the nth 1,000 items and concatenating to the first page. If so, that matches what I'm seeing in #12978 where the Pulse Slack list is limited to 2,000 items (presumably 1,000 channels and 1,000 users).

Reply:
Yup, I put a limit in there of 1000 items as a failsafe to make sure we didn't get into an infinite loop. Previously we only fetched the first page (so only the first 100) results. 

It looks like the limit is too small in reality so we can bump that limit. Unfortunately it doesn't look like the Slack API has a way to search for channels by name so it's not great that we'd have to fetch potentially hundreds of channels to find the `metabase_files` channel